### PR TITLE
[Instance Assignment] Add configs for instance assignment

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
- Also enhance JsonUtils to support TypeReference for deserialization

### CODE REVIEW

File: pinot-common/src/main/java/org/apache/pinot/common/config/TableConfig.java [hunk 1]
```diff
@@ -82,6 +85,10 @@
   @NestedConfig
   private RoutingConfig _routingConfig;
 
+  @ConfigKey(INSTANCE_ASSIGNMENT_CONFIG_MAP_KEY)
+  @UseChildKeyHandler(InstanceAssignmentConfigMapChildKeyHandler.class)
+  private Map<InstancePartitionsType, InstanceAssignmentConfig> _instanceAssignmentConfigMap;
```
Comment (snleee) [inline pinot-common/src/main/java/org/apache/pinot/common/config/TableConfig.java]: Map will be the best approach? We should consider other options to make this simpler. (e.g. put `instanceAssignmentConfigMap` + override config. `TenantConfig` is also using overriding config approach.

Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/config/TableConfig.java]: After consideration, I think explicitly putting the type (OFFLINE/CONSUMING/COMPLETED) is more clear than using the override. For realtime segments, if COMPLETED config does not exist, use CONSUMING config instead

Comment (snleee) [reply pinot-common/src/main/java/org/apache/pinot/common/config/TableConfig.java]: In that case, do you think that we need to wrap those 3 config into some other class? It seems that adding 3 similar configs to the top level table config seems not too clean.

File: pinot-common/src/main/java/org/apache/pinot/common/config/TagNameUtils.java [hunk 2]
```diff
@@ -101,4 +102,18 @@ public static String getTenantNameFromTag(String tag) {
     }
     return tag;
   }
+
+  public static String getServerTagFromTableConfigAndInstancePartitionsType(TableConfig tableConfig,
+      InstancePartitionsType instancePartitionsType) {
+    switch (instancePartitionsType) {
+      case OFFLINE:
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/org/apache/pinot/common/config/TagNameUtils.java]: How does this relate to methods in RealtimeTagConfig class?

Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/config/TagNameUtils.java]: InstancePartitions is associated with server type (OFFLINE/CONSUMING/COMPLETED). Currently that is configured through the tenant config, and we need this method to extract the server tag from the tenant config.

**Pull request merged**